feat(plugin): RelationshipNotifier

lelboatz opened the pull request:
This plugin will send show a notice at the top of the screen (and a desktop notification if discord isn't focused) when a friend removes you or cancels their friend request to you and when you are removed from a group chat or a server.

This also fixes transitionTo not working (credit to the sun from the vencord discord server)

Vendicated commented on src/plugins/relationshipNotifier/utils.ts:
```diff
@@ -0,0 +1,81 @@
+            iconURL: channel.icon ? `https://cdn.discordapp.com/channel-icons/${channel.id}/${channel.icon}.png` : undefined
+        });
+    }
+}
```
you don't seem to persist these anywhere. Why not do that (DataStore is the best way to do that) so you can also notify for changes while Discord was closed?

lelboatz replied:
done

lelboatz commented on src/plugins/relationshipNotifier/utils.ts:
```diff
@@ -0,0 +1,81 @@
+import { NavigationRouter } from "@webpack/common";
+import { Channel, Guild } from "discord-types/general";
+
+import relationshipNotifier from "./index";
```
fixed

lelboatz commented:
i fixed the eslint issues, can you run the workflow again?

lelboatz commented on src/webpack/common.tsx:
```diff
@@ -152,9 +153,9 @@ export const Clipboard = mapMangledModuleLazy('document.queryCommandEnabled("cop
-    transitionTo: filters.byCode('"transitionTo '),
-    transitionToGuild: filters.byCode('"transitionToGuild'),
+export const NavigationRouter = mapMangledModuleLazy("Transitioning to external path", {
+    transitionTo: filters.byCode("Transitioning to external path"),
```
this breaks transitionTo

Vendicated replied:
it's already fixed

lelboatz replied:
oh ok cool

Vendicated commented:
no idea why restore didnt work. In any case i fixed it now

Vendicated commented on src/plugins/relationshipNotifier/settings.ts:
```diff
@@ -0,0 +1,55 @@
+const settings = definePluginSettings({
+    notifications: {
+        type: OptionType.BOOLEAN,
+        description: "Send a desktop notification discord is unfocused",
```
```suggestion
        description: "Also send a desktop notification when discord is unfocused",
```

Vendicated commented on src/plugins/relationshipNotifier/settings.ts:
```diff
@@ -0,0 +1,55 @@
+import { definePluginSettings } from "@api/settings";
+import { OptionType } from "@utils/types";
+
+const settings = definePluginSettings({
```
why not just ```suggestion
export default definePluginSettings({
```

Vendicated commented:
seems to work pretty well but the top banners are a really bad way to notify for this imo since they lack pfp capabilities and it's just ugh. I've been planning to implement neat in-app notifications for a while now

lelboatz commented:
> seems to work pretty well but the top banners are a really bad way to notify for this imo since they lack pfp capabilities and it's just ugly. I've been planning to implement neat in-app notifications for a while now

i agree it would be nice to have an in-app notification system but the main reason i chose top banners was because people often miss in-app notifications if they have discord open on one screen while doing something else. top banners will stay on the screen until dismissed or discord is restarted. i would however be more than happy to switch to another notification system if it provided some sort of notification log and/or a way to keep them on screen until the user interacts with it.

Vendicated commented:
Hi! I just merged Notifications. They don't currently have a way to not auto dismiss but you could add that if you wanted

lelboatz commented:
> Hi! I just merged Notifications. They don't currently have a way to not auto dismiss but you could add that if you wanted

will add notifications to this soon, as for auto dismiss i'll figure something out for that

lelboatz commented:
implemented the notification api. i also removed the setting which toggles desktop notifications while discord is unfocused since you added that in the api and added an option to show notices at the top of the screen as well (instead of making notifications not auto dismiss)

Vendicated commented:
> i also removed the setting which toggles desktop notifications while discord is unfocused

what?

Vendicated commented:
oh you were talking about the plugin's feature, i thought you removed it from the notificiations api

Vendicated commented on src/plugins/relationshipNotifier/utils.ts:
```diff
@@ -0,0 +1,142 @@
+
+const DMStore = findByPropsLazy("getSortedPrivateChannels");
+const GuildStore = findByPropsLazy("getGuilds", "getGuild");
+const Notifications = findByPropsLazy("showNotification", "requestPermission");
```
unused?

Vendicated commented on src/plugins/relationshipNotifier/utils.ts:
```diff
@@ -0,0 +1,142 @@
+import SimpleGuild from "./types/SimpleGuild";
+
+const DMStore = findByPropsLazy("getSortedPrivateChannels");
+const GuildStore = findByPropsLazy("getGuilds", "getGuild");
```
this is in webpack commons!

Vendicated commented on src/plugins/relationshipNotifier/utils.ts:
```diff
@@ -0,0 +1,142 @@
+export async function syncAndRunChecks() {
+    const oldGuilds: Map<string, SimpleGuild> | undefined = await DataStore.get("relationship-notifier-guilds");
+    const oldGroups: Map<string, SimpleGroupChannel> | undefined = await DataStore.get("relationship-notifier-groups");
+    const oldFriends: { friends: string[]; requests: string[] } | undefined = await DataStore.get("relationship-notifier-friends");
```
you could Promise.all these
```ts
const [oldGuilds, oldGroups, oldFriends] = await Promise.all([
    DataStore.get("relationship-notifier-guilds"),
    DataStore.get("relationship-notifier-groups"),
    DataStore.get("relationship-notifier-friends")
]);
```

Vendicated replied:
actually there's DataStore.getMany

lelboatz replied:
how do i type the vars with DataStore.getMany

Vendicated replied:
Sorry I forgot to reply.

```ts
const [oldGuilds, oldGroups, oldFriends] = await Promise.all([
    DataStore.get("relationship-notifier-guilds"),
    DataStore.get("relationship-notifier-groups"),
    DataStore.get("relationship-notifier-friends")
]) as [Map<string, SimpleGuild>, Map<string, SimpleGroupChannel>, Record<"friends" | "requests", string[]>];
```

lelboatz commented:
this is ready to be merged btw (unless someone wants to review some more)

Vendicated commented:
Since the code wasn't really matching the style used by the rest of the project, I rewrote / cleaned up some parts
